used-before-assignment not flagged for import while triggering UnboundLocalError

### Steps to reproduce

`importExample.py`
```python
# pylint: disable=missing-docstring,invalid-name
import time

def test():
    print(time.time())
    import time
    print(time.time())

test()
```
`variableExample.py`
```python
# pylint: disable=missing-docstring,invalid-name
A = 1

def test():
    print(A)
    A = 2
    print(A)

test()
```



### Current behavior

The execution of these two scripts lead to a similar outcome:

```
$ python importExample.py 
Traceback (most recent call last):
  File "importExample.py", line 9, in <module>
    test()
  File "importExample.py", line 5, in test
    print(time.time())
UnboundLocalError: local variable 'time' referenced before assignment
```

```
$ python variableExample.py 
Traceback (most recent call last):
  File "variableExample.py", line 9, in <module>
    test()
  File "variableExample.py", line 5, in test
    print(A)
UnboundLocalError: local variable 'A' referenced before assignment
```
`pylint` detects the problem correctly for the variable, but not for the import

```
$ pylint variableExample.py 
************* Module variableExample
variableExample.py:6:4: W0621: Redefining name 'A' from outer scope (line 2) (redefined-outer-name)
variableExample.py:5:10: E0601: Using variable 'A' before assignment (used-before-assignment)

-----------------------------------
Your code has been rated at 0.00/10
```

```
$ pylint importExample.py 
************* Module importExample
importExample.py:6:4: W0621: Redefining name 'time' from outer scope (line 2) (redefined-outer-name)
importExample.py:6:4: W0404: Reimport 'time' (imported line 2) (reimported)
importExample.py:6:4: C0415: Import outside toplevel (time) (import-outside-toplevel)
importExample.py:2:0: W0611: Unused import time (unused-import)

------------------------------------------------------------------
Your code has been rated at 3.33/10 (previous run: 1.67/10, +1.67)
```
### Expected behavior

I'd expect to see `E0601` raised in the case of the import as well. 

### pylint --version output

```
$ pylint --version
pylint 3.0.0a3
astroid 2.6.0
Python 3.9.5 | packaged by conda-forge | (default, Jun 19 2021, 00:32:32) 
[GCC 9.3.0]
```

--- a/doc/whatsnew/2/2.15/index.rst
+++ b/doc/whatsnew/2/2.15/index.rst
@@ -46,6 +46,10 @@ False negatives fixed
 
   Closes #5653
 
+* Emit ``used-before-assignment`` when calling nested functions before assignment.
+
+  Closes #6812
+
 
 Other bug fixes
 ===============

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -204,11 +204,12 @@ def _detect_global_scope(node, frame, defframe):
         return False
     if isinstance(frame, nodes.FunctionDef):
         # If the parent of the current node is a
-        # function, then it can be under its scope
-        # (defined in, which doesn't concern us) or
+        # function, then it can be under its scope (defined in); or
         # the `->` part of annotations. The same goes
         # for annotations of function arguments, they'll have
         # their parent the Arguments node.
+        if frame.parent_of(defframe):
+            return node.lineno < defframe.lineno
         if not isinstance(node.parent, (nodes.FunctionDef, nodes.Arguments)):
             return False
     elif any(

--- a/tests/functional/u/used/used_before_assignment.py
+++ b/tests/functional/u/used/used_before_assignment.py
@@ -6,3 +6,10 @@ __revision__ = None
 MSG = "hello %s" % MSG  # [used-before-assignment]
 
 MSG2 = "hello %s" % MSG2  # [used-before-assignment]
+
+def outer():
+    inner()  # [used-before-assignment]
+    def inner():
+        pass
+
+outer()

--- a/tests/functional/u/used/used_before_assignment.txt
+++ b/tests/functional/u/used/used_before_assignment.txt
@@ -1,2 +1,3 @@
 used-before-assignment:6:19:6:22::Using variable 'MSG' before assignment:HIGH
 used-before-assignment:8:20:8:24::Using variable 'MSG2' before assignment:HIGH
+used-before-assignment:11:4:11:9:outer:Using variable 'inner' before assignment:HIGH

